rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/rest/TbHttpClientTest.java

```diff
@@ -58,4 +82,63 @@ public void givenNull_whenGetEventLoop_ThenReturnShared() {
         eventLoop = client.getSharedOrCreateEventLoopGroup(null);
         assertThat(eventLoop, instanceOf(NioEventLoopGroup.class));
     }
+
+    @Test
+    public void testProcessMessageWithJsonInUrlVariable() throws Exception {
+        var config = new TbRestApiCallNodeConfiguration()
+                .defaultConfiguration();
+        config.setRequestMethod(GET_METHOD);
+        config.setRestEndpointUrlPattern(ENDPOINT_URL);
+        config.setUseSimpleClientHttpFactory(true);
+
+        var asyncRestTemplate = mock(AsyncRestTemplate.class);
+        var uriCaptor = ArgumentCaptor.forClass(URI.class);
+
+        var responseEntity = new ResponseEntity<>(
+                "{}",
+                new HttpHeaders(),
+                HttpStatus.OK
+        );
+
+        when(asyncRestTemplate.exchange(
+                uriCaptor.capture(),
+                any(),
+                any(),
+                eq(String.class)
+        )).thenReturn(new AsyncResult<>(responseEntity));
+
+        var httpClient = new TbHttpClient(config, eventLoop);
+        httpClient.setHttpClient(asyncRestTemplate);
+
+        var msg = TbMsg.newMsg(
+                "Main", "GET", new DeviceId(EntityId.NULL_UUID),
+                TbMsgMetaData.EMPTY, "{}"
+        );
+        var successMsg = TbMsg.newMsg(
+                "SUCCESS", msg.getOriginator(),
+                msg.getMetaData(), msg.getData()
+        );
+
+        var ctx = mock(TbContext.class);
+        when(ctx.transformMsg(
+                        eq(msg), eq(msg.getType()),
+                        eq(msg.getOriginator()),
+                        eq(msg.getMetaData()),
+                        eq(msg.getData())
+                )).thenReturn(successMsg);
+
+        httpClient.processMessage(ctx, msg);
+
+        verify(ctx, times(1)).transformMsg(
+                eq(msg), eq(msg.getType()),
+                eq(msg.getOriginator()),
+                eq(msg.getMetaData()),
+                eq(msg.getData())
+        );
+        verify(ctx, times(1))
+                .tellSuccess(eq(successMsg));
+
+        URI uri = UriComponentsBuilder.fromUriString(ENDPOINT_URL).build().encode().toUri();
```

Comment:
I don't think this is a good test. It is better to create some kind of a dummy server that listens to all requests and echoes the URI in response. Otherwise, you are not testing the real behavior of the component. Also, this test does not check behavior for "normal" URLs

Reply:
Other option is to replace the dummy server with the:
1) extract your changes ( " URI uri = UriComponentsBuilder.fromUriString(endpointUrl).build().encode().toUri();") to the new method
2) cover the method with unit tests with all sorts of input URL templates
3) make sure your method was executed in the main message processing flow.